More robust point cloud post processor

AnimatedRNG opened the pull request:
Still don't have the "change previous stages so that they write out ES coordinates to the texture" part working; everything else looks okay though.

pjcozzi commented on Source/Scene/PointCloudPostProcessor.js:
```diff
@@ -72,20 +78,29 @@ define([
+            width : screenWidth,
+            height : screenHeight,
+            pixelFormat : PixelFormat.RGBA,
+            pixelDatatype : PixelDatatype.FLOAT,
```
Do we need to check for an extension for this?

WebGL 2 or https://www.khronos.org/registry/webgl/extensions/OES_texture_float/

pjcozzi commented on Source/Scene/PointCloudPostProcessor.js:
```diff
@@ -72,20 +78,29 @@ define([
-                sampler : createSampler()
-            });
+        for (i = 0; i < 3; ++i) {
+            if (i < 2) {
```
Use two separate `for` loops; it will be cleaner.

pjcozzi commented on Source/Scene/PointCloudPostProcessor.js:
```diff
@@ -95,65 +110,138 @@ define([
-            // TODO : for now assume than any pass can write depth, possibly through EXT_frag_depth.
-            // Definitely needed for the initial render into the FBO, possibly also the ping-pong processing.
-            framebuffers[i] = new Framebuffer({
+        /* We want to reuse textures as much as possible, so here's the order
```
Stick with the `//` comment style.

This is actually not in the Coding Guide, but I'm pretty sure it is a common practice throughout Cesium except for comments for generating reference doc with JSDoc.

Could you open a separate PR to mention this at the bottom of the [Basic Code Construction](https://github.com/AnalyticalGraphicsInc/cesium/tree/master/Documentation/Contributors/CodingGuide#basic-code-construction)?

pjcozzi commented on Source/Scene/PointCloudPostProcessor.js:
```diff
@@ -95,65 +110,138 @@ define([
+         */
+
+        processor._framebuffers = {
+            "prior": new Framebuffer({
```
Here and similar properties below, `"prior"` -> `prior`

Quotes are not needed here.  When they are, use single quotes to be consistent with the rest of the code base.

pjcozzi commented on Source/Scene/PointCloudPostProcessor.js:
```diff
@@ -72,20 +78,29 @@ define([
-        var framebuffers = new Array(2);
+        var depthTextures = new Array(3);
+
+        var ecTexture = new Texture({
```
Somewhere, we should comment (or rename this) about what "EC" stands for.

pjcozzi commented on Source/Scene/PointCloudPostProcessor.js:
```diff
@@ -170,49 +258,54 @@ define([
+        var framebuffers = processor._framebuffers;
+        var clearCommands = new Array(4);
+        i = 0;
+        for (var name in framebuffers) {
```
Throughout, it looks like `_framebuffers` is used as both an array and object.  Be consistent.  Perhaps this can be an array with constants in an enum for look ups.

pjcozzi commented on Source/Scene/PointCloudPostProcessor.js:
```diff
@@ -222,15 +315,59 @@ define([
             destroyFramebuffers(processor);
-            createFramebuffers(processor, context);
-            updateCommandFramebuffers(processor);
+            //createFramebuffers(processor, context);
```
We generally don't commit code that is commented out.  Usually, remove it.

Otherwise, it can stay in your branch like this, but we would remove it before merging into `master`.

pjcozzi commented on Source/Scene/PointCloudPostProcessor.js:
```diff
@@ -222,15 +315,59 @@ define([
+            'void main() \n' +
+            '{ \n' +
+            '    czm_point_cloud_post_process_main(); \n' +
+            //'    gl_FragData[0] = gl_FragColor;' +
```
Same comment.

AnimatedRNG commented:
Should be ready to review @lilleyse

pjcozzi commented:
@lilleyse I'd like to review this before merging, but please do the first review and iterations and then @mention when ready me so I can take a quick pass.

lilleyse commented on Source/Renderer/GLSLModernizer.js:
```diff
@@ -0,0 +1,351 @@
+        var outputDeclarationRegex = /#define OUTPUT_DECLARATION/;
+        var splitSource = source.split('\n');
+
+        if (source.search(/#version 300 es/g) !== -1) {
```
Use `.test` instead, `search` is not supported in all browsers.

lilleyse commented on Source/Renderer/GLSLModernizer.js:
```diff
@@ -0,0 +1,351 @@
+        
+        var outputDeclarationLine = -1;
+        var number;
+        for (number = 0; number < splitSource.length; number++) {
```
Same `length` comment, make sure to fix throughout.

lilleyse commented on Source/Renderer/GLSLModernizer.js:
```diff
@@ -0,0 +1,316 @@
+    function glslModernizeShaderSource(shaderSource, isFragmentShader) {
+        for (var i = 0; i < shaderSource.sources.length; i++) {
+            shaderSource.sources[i] =
+                glslModernizeShaderText(shaderSource.sources[i], isFragmentShader, i === 0);
```
Similar comment throughout, as long as the line is not excessively long.

AnimatedRNG replied:
How long is too long?

lilleyse replied:
I usually max out at around 120 - 140

lilleyse commented on Source/Renderer/Context.js:
```diff
@@ -268,13 +268,17 @@ define([
 
+        if (this._webgl2) {
+            this._colorBufferFloat = !!getExtension(gl, ['EXT_color_buffer_float']);
+        }
```
Actually @bagnell's suggestion is better: https://github.com/AnalyticalGraphicsInc/cesium/pull/5568#discussion_r124891493

AnimatedRNG commented on LICENSE.md:
```diff
@@ -647,6 +647,28 @@ THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
+### deck.gl
```
The point occlusion shader uses 64-bit primitives internally if the USE_TRIANGLE flag is enabled

AnimatedRNG commented on Source/Renderer/Texture.js:
```diff
@@ -71,6 +71,20 @@ define([
+                    break;
+                case PixelFormat.R:
+                    internalFormat = WebGLConstants.R32F;
+                    break;
```
Good point. I accidentally only included the color-renderable formats in here.

AnimatedRNG commented on Apps/Sandcastle/gallery/3D Tiles Point Cloud Post Processing.html:
```diff
@@ -0,0 +1,166 @@
+var inspectorViewModel = viewer.cesium3DTilesInspector.viewModel;
+inspectorViewModel.picking = false;
+
+viewer.clock.currentTime = new Cesium.JulianDate(2457522.154792);
```
Are you referring to the `currentTime` or the picking?

lilleyse replied:
`currentTime`

AnimatedRNG commented on Source/Scene/Cesium3DTileset.js:
```diff
@@ -1565,6 +1592,25 @@ define([
+        if (addedCommandsLength > 0) {
+            // TODO : only do this if the tileset is purely point clouds.
+            // TODO : this may not work well if point cloud spans multiple frustums
+            // TODO : make the processor a static class so it can be used by multiple tilesets?
```
Regarding the second comment, we currently have multifrustrum support in the code

lilleyse replied:
Ah good - I forgot about that.

mramato commented:
Just an FYI, but I'm getting the following error trying to run this on my machine:

![image](https://user-images.githubusercontent.com/1495196/28542953-2823061e-708c-11e7-92e8-75b54b717596.png)

Windows 10, Chrome, Nvidia 650ti

lilleyse commented:
@AnimatedRNG is `point-cloud-processor-ss-optimized` far enough along to be merged into here now?

@mramato the fix for that shader bug is on the `-optimized` branch.

mramato commented:
Thanks @lilleyse I'm still getting a different `Fragment shader failed to compile.` error using `point-cloud-processor-ss-optimized`, but unfortunately no further information was provided.  Here's the error:

```
[Cesium WebGL] Fragment shader compile log: 
[Cesium WebGL] Fragment shader translation failed.
An error occurred while rendering.  Rendering has stopped.
undefined
RuntimeError: Fragment shader failed to compile.  Compile log: 
Error
    at new RuntimeError (http://localhost:8080/Source/Core/RuntimeError.js:42:19)
    at createAndLinkProgram (http://localhost:8080/Source/Renderer/ShaderProgram.js:232:23)
```

I'll just leave this PR alone until it is further along.  I'm hoping to incorporate it into some demo code I'm writing, so please bump this when it's ready for wider testing.  Thanks.

AnimatedRNG commented:
Hi @mramato I merged in the fixes from `point-cloud-processor-ss-optimized`. Things should work now. If you're still getting issues from fragment shaders not compiling, I'd be happy to take a look at the log.

mramato commented:
> If you're still getting issues from fragment shaders not compiling, I'd be happy to take a look at the log.

Thanks, I finally got it to work, but most of the time it still crashes on startup with a "width must be greater than 0" exception.  (at which point Chrome treats it like a webgl crash).  If I just keep refreshing the page, it works eventually.  Let me know if you want to debug anything on my machine.

AnimatedRNG commented:
@mramato Still haven't fixed the crash on load issue, but now the post processor is off by default. When you turn it on, it should work fine because the terrain isn't loading.

The last few commits add support for "random reduction". This feature will randomly reduce the point splat size for the point arrays used by the `SectorHistogram`, `EdgeCulling`, and `DensityEstimation` passes, based on the parameter `randomReductionFactor`. Randomly reducing the point splat size reduces the number of generated fragments, at the cost of randomly failing to convey neighbor information. A reduction factor of 1.0 means that reduction is not taking place (maximum quality), whereas a reduction factor of 0.0 means that the point splat size will be zero (maximum performance).

You can use this parameter to find a nice spot on the performance/quality curve.

AnimatedRNG commented:
Regarding my last comment; I've removed `randomReductionFactor` because I've also removed point arrays. Turns out that performing a 40x more texture reads is more performant than using min-blending.

Most of the crashes that occurred on Windows should be fixed now, as well as the grid pattern artifacts.

AnimatedRNG commented:
While `randomReductionFactor` has been removed, and equivalent feature `dropoutFactor`, has been added. Unlike `randomReductionFactor`, it requires WebGL 2 and only yields acceptable savings on extreme settings. However, it *does* give us the performance vs quality slider that we want.

mramato commented:
@AnimatedRNG sorry if you already know about this, but the mesh visualization doesn't work if a billboard is onscreen.  You can reproduce by adding the following snippet to your new Sandcastle example:

```js
    viewer.entities.add({
        position : new Cesium.Cartesian3(-74391848.22441247,19357428.39048527,84933455.20184723),
        billboard : {
            image : '../images/Cesium_Logo_overlay.png'
        }
    });
```

1. From the default view, tile the camera to view the horizon and you'll see the billboard off in the distance.
2. Enable surface visualization and the entire point cloud will simply disappear.
3. Change rotate the camera and as soon as the billboard is offscreen and culled, the point cloud will re-appear to the surface working properly.

AnimatedRNG commented:
Blending should be fixed -- eye-space coordinates are now used internally so there's no need to clamp things. The race condition that popped up on load should also be fixed @mramato .

I still have to port over some performance stuff (see 1820dc2a3d5079f38eb53eb9cc74550249a7e0e9) to the WebGL 2 shader; will do that soon.

@mramato I'll try your billboard test on my end and see what happens. I suspect your issue is related to depth blending, but I'll know for sure in a bit.

AnimatedRNG commented:
@mramato I tried the billboard test and everything seems to work now (i.e I can see the filtered point cloud and the billboard at the same time)

mramato commented:
Awesome, thanks.

lilleyse commented:
On my machine I'm getting strange results for Depth View and AO View in the Sandcastle demo:

<img width="256" alt="depth-view" src="https://user-images.githubusercontent.com/915398/29126729-286e2f5a-7ced-11e7-994a-a8f8f9971c5d.png">
<img width="248" alt="ao-view" src="https://user-images.githubusercontent.com/915398/29126722-254c403c-7ced-11e7-9b45-3acfa8c45286.png">

lilleyse commented on LICENSE.md:
```diff
@@ -672,6 +672,28 @@ https://github.com/KhronosGroup/glTF-WebGL-PBR
 >CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE
 >OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 
+### deck.gl
```
Add a link to the repo and use quoted text for the license text, similar to the others.

lilleyse commented on Source/Scene/Cesium3DTileset.js:
```diff
@@ -1501,6 +1557,11 @@ define([
         statistics.numberOfCommands = (commandList.length - numberOfInitialCommands);
 
+        if (addedCommandsLength > 0) {
+            // TODO : only do this if the tileset is purely point clouds.
```
Is this TODO needed anymore?

Also just to make it clear that the processor only runs when it's enabled, the check could be
`if (tileset.pointCloudPostProcessorOptions.enabled && (addedCommandLength > 0))`

lilleyse commented on Source/Scene/PointCloud3DTileContent.js:
```diff
@@ -898,6 +913,7 @@ define([
 
         var vs = 'attribute vec3 a_position; \n' +
                  'varying vec4 v_color; \n' +
+                 'uniform float u_pointAttenuationMaxSize; \n' +
```
This can be packed with `u_pointSizeAndTilesetTime`

AnimatedRNG replied:
That might be a little bit hard to read -- `u_pointSizeAndTilesetTimeAndAttenuation` maybe?

lilleyse replied:
Sure.

AnimatedRNG replied:
I think it might be a bit easier to keep `u_pointSizeAndTilesetTime` as is because the `PointCloudPostProcessor` directly edits the uniform map, replacing the `u_pointAttenuationMaxSize` uniform function with a function that just returns the attenuation bound. It can't edit `u_pointSizeAndTilesetTimeAndAttenuation` as cleanly because of the other two attributes (point size and tileset time) that are returned by the function.

lilleyse commented on Source/Scene/PointCloud3DTileContent.js:
```diff
@@ -1004,7 +1020,13 @@ define([
+              '    float attenuationFactor = \n' +
+              '    ((position_EC.z - czm_clampedFrustum.x) /\n' +
+              '    (czm_clampedFrustum.y - czm_clampedFrustum.x)); \n' +
+              '    gl_PointSize *= mix(u_pointAttenuationMaxSize, 1.0, attenuationFactor); \n\n' +
```
The point attenuation code should be added to the shader only if `_pointAttenuationMaxSize > _pointSize`. Also point attenuation should probably not be done at all if the style has a `pointSize` property.

lilleyse commented on Source/Scene/PointCloud3DTileContent.js:
```diff
@@ -255,6 +258,15 @@ define([
+         * @private
+         */
+        pointAttenuationMaxSize : {
+            get : function() {
```
Is this getter used anywhere? If not, remove.

lilleyse commented on Source/Scene/PointCloud3DTileContent.js:
```diff
@@ -113,6 +113,9 @@ define([
         this._hasBatchIds = false;
 
+        // Used to determine how to attenuate points
+        this._pointAttenuationMaxSize = 1.0;
```
Place this near `pointSize` since they are a common concept.

lilleyse commented on Source/Scene/PointCloud3DTileContent.js:
```diff
@@ -1004,7 +1020,13 @@ define([
         }
 
-        vs += '    color = color * u_highlightColor; \n';
+        vs += '    vec4 position_EC = czm_view * vec4(position_absolute, 1.0); \n' +
```
For style purposes rename `position_EC` to `positionEC`. It may be worth renaming `position_absolution` to `positionWC` (except line 784 which relies on the name being the lowercase of the semantic name).

AnimatedRNG commented:
@lilleyse The depth view is (probably perpetually) broken because of the new eye-space coordinates code. Right now it's just visualizing the eye-space coordinates, which I figured was useful enough to leave in. I just fixed the AO view.

lilleyse commented on Source/Shaders/PostProcessFilters/PointOcclusionPassGL1.glsl:
```diff
@@ -0,0 +1,219 @@
+            // from the viewer to the center and the vector to the horizon pixel.
+            // We normalize both vectors first because we only care about their relative
+            // directions
+            // TODO: Redo the math and figure out whether the result should be negated or not
```
Address this TODO.

lilleyse commented on Specs/Scene/PointCloudPostProcessorSpec.js:
```diff
@@ -0,0 +1,85 @@
+            scene.renderForSpecs();
+            var originalLength = scene.frameState.commandList.length;
+
+
```
Remove empty line.

lilleyse commented on Specs/Scene/PointCloudPostProcessorSpec.js:
```diff
@@ -0,0 +1,85 @@
+        PointCloud3DTileContent,
+        Cesium3DTilesTester,
+        createScene,
+        when) {
```
A lot of these includes are not used.

lilleyse commented on Specs/Scene/PointCloudPostProcessorSpec.js:
```diff
@@ -0,0 +1,85 @@
+        var camera = scene.camera;
+        camera.frustum = new PerspectiveFrustum();
+        camera.frustum.aspectRatio = scene.drawingBufferWidth / scene.drawingBufferHeight;
+        camera.frustum.fov = CesiumMath.toRadians(60.0);
```
Is the camera setup needed?
`scene.morphTo3D(0.0);` can also be removed.

AnimatedRNG commented on Source/Shaders/PostProcessFilters/PointCloudPostProcessorBlendPass.glsl:
```diff
@@ -0,0 +1,112 @@
+    } else {
+        float depth = czm_eyeToWindowCoordinates(vec4(ec.xyz, 1.0)).z;
+        gl_FragColor = color;
+        gl_FragDepthEXT = depth;
```
Yeah, tried @mramato's billboard test and it seemed to work. Also tried testing against the terrain itself.

AnimatedRNG commented on Source/Scene/PointCloudPostProcessor.js:
```diff
@@ -0,0 +1,1108 @@
+        this.dropoutFactor = options.dropoutFactor;
+        this.delay = options.delay;
+
+        this._pointArray = undefined;
```
As far as I'm aware, I don't think there's anything else.

AnimatedRNG commented on Source/Scene/PointCloudPostProcessor.js:
```diff
@@ -0,0 +1,1108 @@
+        var screenHeight = context.drawingBufferHeight;
+
+        var colorTextures = new Array(2);
+        var depthTextures = new Array(3);
```
Good catch. It's a bit of old code I never cleaned up

AnimatedRNG commented on Source/Scene/PointCloudPostProcessor.js:
```diff
@@ -0,0 +1,1109 @@
+                sampler : createSampler()
+            });
+
+            depthTextures[i] = new Texture({
```
Technically all the textures are eye-space textures, so they should all be named `ECTextures` or something. With a bit of a refactor, we can remove the original `ECTexture` altogether.

AnimatedRNG commented on Source/Scene/PointCloudPostProcessor.js:
```diff
@@ -0,0 +1,1109 @@
+        // There used to be an explanation of how this worked here
+        // but it got too long.
+        // TODO: Find a better place to put an explanation of what all
+        // the framebuffers are meant for.
```
I'm going to leave this explanation out of the file entirely, as discussed

AnimatedRNG commented on Source/Scene/PointCloudPostProcessor.js:
```diff
@@ -0,0 +1,1109 @@
+            pointCloud_depthTexture : function() {
+                return processor._depthTextures[0];
+            },
+            neighborhoodVectorSize : function() {
```
The original post processing framework that you wrote didn't have a `u_` prefix before the samplers -- should I keep it that way?

lilleyse replied:
Probably just an oversight on my part, everything can be `u_`.

AnimatedRNG replied:
Okay fixed it

AnimatedRNG commented on Source/Scene/PointCloudPostProcessor.js:
```diff
@@ -0,0 +1,1109 @@
+                );
+
+                // TODO: Even if the filter is disabled,
+                // point attenuation settings are not! Fix this behavior.
```
I don't think there is an easy way to avoid this issue. Do you have any ideas?

lilleyse replied:
If the point cloud processor is disabled it should use the original, unmodified uniform map. Related to this comment: https://github.com/AnalyticalGraphicsInc/cesium/pull/5455#discussion_r132297181 -

> This area is also in need of cloning. The combine function is another option that we often use when combining different uniform maps together.

AnimatedRNG replied:
Okay, should be fixed

AnimatedRNG commented on Source/Scene/PointCloudPostProcessor.js:
```diff
@@ -0,0 +1,1109 @@
+                derivedCommand.uniformMap = derivedCommandUniformMap;
+
+                derivedCommand.pass = Pass.CESIUM_3D_TILE; // Overrides translucent commands
+                command.dirty = false;
```
The `command.dirty` line?

AnimatedRNG commented on Source/Scene/PointCloudPostProcessor.js:
```diff
@@ -0,0 +1,1109 @@
+
+    /**
+     * Part of the {@link Cesium3DTileContent} interface.
+     */
```
I guess I'll just inherit documentation then?

AnimatedRNG replied:
Actually, I'll just copy the docs over

AnimatedRNG commented on Source/Scene/PointCloudPostProcessor.js:
```diff
@@ -0,0 +1,1109 @@
+            '#extension GL_EXT_draw_buffers : enable \n' +
+            '#define densityView \n' +
+            '#define densityScaleFactor 10.0 \n' +
+            '#define EPS 1e-6 \n' +
```
There's a couple areas where smaller epsilons are needed (i.e 1e-9, but cesium doesn't have an built-in epsilon that small); for now I'm going to keep those epsilons.

AnimatedRNG commented on Source/Scene/PointCloudPostProcessor.js:
```diff
@@ -0,0 +1,1109 @@
+        processor._dirty = dirty;
+    }
+
+    function replaceConstants(sourceStr, constantName, replacement) {
```
So, just to clarify, you're not suggesting using ShaderSource's `defines` array, and instead constructing the shader string on the fly?

lilleyse replied:
Normally I would say add them to the `defines` array, but since they are setting constants it's different than what `defines` is traditionally used for. So yeah, adding the defines to the shader string is the right approach for now.

AnimatedRNG replied:
The problem with just appending the `#define`s to the shader string is that it's not clear where the new lines should be appended. They can't be appended at the start because [GLSL requires](https://www.khronos.org/opengl/wiki/Core_Language_(GLSL)#Extensions) `#extension`s to be listed right after the version declaration (i.e so they have to be appended after the extension declaration(s) but before any uniform declarations).

Ideally we'd solve this problem by working with `ShaderSource` objects rather than just shader strings. Unfortunately, we also want to use the `createQuadViewportFS` method from `Context`, which doesn't take a ShaderSource.

lilleyse replied:
The `#extensions` shouldn't be a problem as of the GLSL Modernizer PR:  https://github.com/AnalyticalGraphicsInc/cesium/pull/5623

lilleyse commented on Source/Scene/PointCloudPostProcessor.js:
```diff
@@ -0,0 +1,1109 @@
+            this.sigmoidSharpness = tileset.pointCloudPostProcessorOptions.sigmoidSharpness;
+            this.dropoutFactor = tileset.pointCloudPostProcessorOptions.dropoutFactor;
+            this.delay = tileset.pointCloudPostProcessorOptions.delay;
+            dirty = true;
```
@AnimatedRNG did you see this comment?

lilleyse commented on Source/Scene/PointCloudPostProcessor.js:
```diff
@@ -0,0 +1,1070 @@
+            commandList.push(debugViewCommand);
+        }
+
+        commandList.push(clearCommands['prior']);
```
Style - `clearCommands.prior`

AnimatedRNG commented on Source/Scene/PointCloud3DTileContent.js:
```diff
@@ -766,7 +770,7 @@ define([
 
         // Edit the function header to accept the point position, color, and normal
-        return source.replace('()', '(vec3 position, vec3 position_absolute, vec4 color, vec3 normal)');
+        return source.replace('()', '(vec3 position, vec3 positionWC, vec4 color, vec3 normal)');
```
I'm not exactly sure if the latest commit fixes this, but I think it's fixed

lilleyse commented on Source/Scene/PointCloudPostProcessor.js:
```diff
@@ -0,0 +1,1178 @@
+                derivedCommand.castShadows = false;
+                derivedCommand.receiveShadows = false;
+
+                var derivedCommandRenderState = RenderState.fromCache(derivedCommand.renderState);
```
Do a `clone` here instead since the `RenderState.fromCache` call is below. For example, look at https://github.com/AnalyticalGraphicsInc/cesium/blob/master/Source/Scene/Cesium3DTileBatchTable.js#L1329-L1332.

lilleyse commented on Source/Scene/PointCloudPostProcessor.js:
```diff
@@ -0,0 +1,1178 @@
+        processor._stencilMaskTexture = stencilMaskTexture;
+    }
+
+    function addConstants(sourceFS, constantName, replacement) {
```
Rename `replacement` to `value`.

lilleyse commented on Source/Scene/PointCloudPostProcessor.js:
```diff
@@ -0,0 +1,1178 @@
+        this._drawCommands = undefined;
+        this._clearCommands = undefined;
+
+        this.densityScaleFactor = '10.0';
```
This can be a number instead of a string.

AnimatedRNG replied:
It's slightly easier for it to remain a string because javascript keeps representing the number `10.0` as `10` when it is stringified. When the preprocessor define gets copied in, there are casting issues.

lilleyse commented on Specs/Scene/PointCloudPostProcessorSpec.js:
```diff
@@ -0,0 +1,74 @@
+        scene.primitives.removeAll();
+    });
+
+    it('enabling the point cloud post processor increases the number of draw calls', function() {
```
This test fails on Travis since the WebGL Stub does not enable the required extensions for the processor. You may want to make `processingSupported` in `PointCloudPostProcessor` a static function like `PointCloudPostProcessor.isSupported` and check that before running the test.

lilleyse commented on Source/Scene/PointCloudPostProcessor.js:
```diff
@@ -0,0 +1,1178 @@
+        densityEdgeCullFS = addConstants(
+            densityEdgeCullFS,
+            'epsilon8',
+            1e-8
```
Thinking about these defines some more, it might be cleaner to keep unchanging defines like this in the shader code, and only call `addConstants` on those that do change.

lilleyse commented:
@pjcozzi do you want to review this before it goes into `point-cloud-processor`?

pjcozzi commented:
> do you want to review this...

Yes

> ...before it goes into point-cloud-processor?

No

I'll review as soon as I can - and comment here - but please merge in the meantime.

lilleyse commented:
Thanks for all the work on this @AnimatedRNG!
